service/controller/chart/resource/status/create.go

```diff
@@ -73,27 +75,32 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 			status = releaseStatusCordoned
 			reason = key.CordonReason(cr)
 		} else {
-			status = releaseContent.Status
-			if releaseContent.Status != helmclient.StatusDeployed {
-				if cc.Status.Release.FailedMaxAttempts {
+			// releaseContent contains the actual status got from the current Helm release
+			status = releaseContent.Status                          // deployed
+			if releaseContent.Status != helmclient.StatusDeployed { // we compare the actual to the desired, and we ignore what is in the controller context
+				if cc.Status.Release.FailedMaxAttempts { // but then we check the controller context here, wut?
 					reason = fmt.Sprintf("Release has failed %d times.\nReason: %s",
 						project.ReleaseFailedMaxAttempts,
 						releaseContent.Description)
 				} else {
-					reason = releaseContent.Description
+					reason = releaseContent.Description // and again we reuse the actual Helm release state
 				}
+			} else if cc.Status.Reason != "" {
+				// It can be that there was previous successful deployment and that is the current status of the Helm release
+				// but a follow-up deployment failed for another version, so we want to set a reason indicating the issue
+				reason = fmt.Sprintf("Deployment failed for %#q with %#q: %#q", cr.Spec.Version, cc.Status.Release.Status, cc.Status.Reason)
 			}
 		}
 	}
 
 	desiredStatus := v1alpha1.ChartStatus{
-		AppVersion: releaseContent.AppVersion,
-		Reason:     reason,
-		Release: v1alpha1.ChartStatusRelease{
-			Revision: to.IntP(releaseContent.Revision),
-			Status:   status,
+		AppVersion: releaseContent.AppVersion, // The desired should depend on the version set in the Chart, not what is deployed by Helm
```

Comment:
There are so many questions here in the comments. Do you want to merge it like that, or would you rather wait for them to get answered?

Reply:
Right, no it is a leftover if making sense of it all.